File: microservices/click/click.js
var AWS = require("aws-sdk");
var express = require('express');
const request = require('request');
const app = express();

var click_event = (anuncio, queryid) => {
    var insertar = [];
    insertar.push(anuncio);
    insertar.push(queryid);
    try{
        request({
            url: "/tracking2",
            method: "POST",
            json: true,
            body: insertar,
        }, function (error, response, body){
            if(error){
                return error;
            }
            console.log(body);
            return "yea";
        });
    }catch(e){
        return e;
    }
}

app.use(express.json());

AWS.config.update({
    "region":"",
    "accessKeyId":"",
    "secretAccessKey":""
});
var docClient = new AWS.DynamoDB.DocumentClient();

app.get('/clickh', function(req, res){
    res.status(200);
    res.send("OK");
});

app.post('/clicki', function(req, res){
    var contenido = req.body;

    var params = {
        TableName: "clicks",
        Item: contenido
    };

    docClient.put(params, function(err, data) {
       if (err) {
           console.log(contenido);
           console.error(err);
           res.status(500);
           res.send("Dynamodb connection error");
       } else {
           console.log("PutItem succeeded");
           res.status(200);
           res.send(contenido);
       }
    });

});

app.get('/click', function(req, res){
    var impression_id = req.query.impression_id;
    var query_id = req.query.query_id;

    if(typeof impression_id != 'undefined' && impression_id && typeof query_id != 'undefined' && query_id){
        var params = {
            TableName : 'clicks',
            Key: {
              'query-id': query_id
            }
        };
        docClient.get(params, function(err, data) {
            if (err) {
              console.log("Error", err);
              res.status(500);
              res.send("Dynamo error");
            } else {
              var siguiente = data.Item['ads'];
              var direccion = "";
              for(var i = 0; i < siguiente.length; i++){
                  if(siguiente[i]['impression_id'] == impression_id){
                      console.log(click_event(siguiente[i], query_id));
                      direccion = siguiente[i]['url'];
                      console.log("hola" , direccion);
                    }
                }
                if(direccion != ""){
                    res.status(302);
                    res.redirect(direccion);
                }else{
                    res.status(404);
                    res.send("Not found");
                }
              }
            });
      }
  });
  
  var server = app.listen(3000, function(){});
  
  

File: microservices/click/click.py
import boto3
import uuid
import json
import datetime
from ast import literal_eval
from flask import Flask
from flask import request, redirect
from decimal import *

app = Flask(__name__)
dynamodb = boto3.resource('dynamodb', region_name=, aws_access_key_id=, aws_secret_access_key=)
table = dynamodb.Table('clicks')

@app.route('/clickh')
def clicks_health():
    return "OK", 200

@app.route('/click')
def clicks_func():
    impression_id = request.args.get('impression_id')
    query_id = request.args.get('query_id')

    if impression_id != None and query_id != None:
        response = table.get_item(Key={'query-id':query_id})
        print(response)
        siguiente = response['Item']['ads']
        direccion = ""
        for i in range(len(siguiente)):
            if siguiente[i]['impression_id'] == impression_id:
                direccion = siguiente[i]['url']
        return redirect(str(direccion), code=302)
    else:
        return "bad", 500

@app.route('/clicki', methods=['POST'])
def click_insert():
    contenido = literal_eval(request.json)
    print(type(contenido))
    table.put_item(
        Item=contenido
    )
    return "OK", 200

app.run(host='0.0.0.0', port=3000)


